Cache API key doc to reduce traffic to the security index

ywangd opened the pull request:
Getting the API key document form the security index is the most time consuing part of the authentication flow (>60% if index is local and >90% if index is remote). This traffic can be avoided with cache. Initially we also plan to avoid deserialising source of GetApiKeyDocument response if "primary term" and "sequence number" are unchanged. However, given the document fetching itself is dominant in execution time, it does not seem to be worthwhile to optimise for deserialisation. Hence, for uncached document, the cache will fetch the document and also simply deserialise it without further checking.

When a large number of API keys are in use, we expect most of them are created with exactly the same role-descriptors and limited-by-role-descriptors. It is wasteful to cache these identical role descriptors repeatitively for different API keys. Hence a two layer caches are put in place:
* The first cache is keyed by docId and values are content of the API key document, except that the role-descriptors and limited-by-role-descriptors are replacecd by their sha256 hashes respectively.
* The second cache is keyed by the above sha256 hashes and values are the corresponding role-descriptors.

The lifecyces of the above two caches are mostly tied together and they need work together in a way to ensure least possibility of stale results. Hence a new `SecurityCaches` class is added and it implements all necessary handlings as discussed in #55836. 

@tvernum This is a quick attempt to implement a cache for API key document. It intentionally omits some parts, e.g. tests, configuration, to get a quick turnaround for proving the idea. It is easier to discuss when we have something concrete. We could try formalise it for v7.9 if feasible. But I am also OK to delay it if more careful evaluation is needed. Thanks!

Resolves: #53940

ywangd committed:
WIP: add cache for API Key document.

ywangd commented:
Please note this PR is built on the yet-to-be-approved #58156. Therefeore many changes are actually not relevant to this PR itself. Here are a list of actual relevant changes:
* [SecurityCaches](https://github.com/elastic/elasticsearch/pull/59376/files#diff-bee5927169b4cdaf17a45067a7e20a38)
* [ApiKeyService](https://github.com/elastic/elasticsearch/pull/59376/files#diff-cc4f9d4c564aae71214ac709ed789b4dR207-R220) and [here](https://github.com/elastic/elasticsearch/pull/59376/files#diff-cc4f9d4c564aae71214ac709ed789b4dR371-R386) and [here](https://github.com/elastic/elasticsearch/pull/59376/files#diff-cc4f9d4c564aae71214ac709ed789b4dR401-R407)
* [Security](https://github.com/elastic/elasticsearch/pull/59376/files#diff-8e35e3eb57150a97e6dcfc4418afc58aR439-R440)

ywangd commented:
@elasticmachine run elasticsearch-ci/1

tvernum commented:
I don't think it's worth pushing this in before FF.
It's a good change, but I think we need time to discuss things such the split doc/auth caches.

ywangd commented:
> I think we need time to discuss things such the split doc/auth caches.
>
Thanks, I agree. The changes are still rough. I also thought about merging doc and auth caches but didn't do it mainly because of complexity and potential implications of complex changes. It requires more time to get right. 
Also a big part of the change is about refactoring the caching behaviour into a separate class, which probably deserves its own PR.

ywangd committed:
Merge remote-tracking branch 'origin/master' into es-53940-apikey-getdoc-avoidance

ywangd committed:
Merge remote-tracking branch 'origin/master' into es-53940-apikey-getdoc-avoidance

ywangd committed:
WIP: Have one cache for API key doc and auth result.

ywangd commented:
I spent some time look into how we could have a single cache for both API key doc and Auth hash. I came up with a [version](https://github.com/elastic/elasticsearch/compare/master...ywangd:es-53940-apikey-getdoc-avoidance-unified-cache) that works. However, it comes with high complexity because the two parts of data have different lifecycles.

With the single cache approach, a cache item contains both content of an API key document and the Authentication hash. These two parts can vary independantly. That is, when the document needs to be refreshed for caching, the authentication hash can still be valid because it is based on user-provided credentials and only associated with the document ID, not its content. Similarly, when the authentication hash needs to be updated, the document can still be valid because unlike authentication hash it does not change based on user-provided credentials. So what we need is a cache where the cache item can be _partially_ invalidated. This requirement is basically asking for two separate caches. It is possible to make it work with a single cache as mentioned earlier, but it leads to high complexity due to the diverging requirement plus how it could work with a distributed system. Therefore, even though I already have an implementation for it, I still prefer the approrach of separate caches. Honestly, I had fun caming with my current implementaiton and trying to catch all edge cases. But I am concerned that it is difficult to explain how it works and maintain it a few years from now.

After discussion with @tvernum, we decided to go with separate caches. There is no overlap between contents of the two caches. Hence we do not have concern in terms of duplication (memory consumption).

ywangd committed:
Revert "WIP: Have one cache for API key doc and auth result."

This reverts commit df2a1d8614a9ac1631e1183ea265fdc7c48d8651.

ywangd committed:
Add security cache common behaviours and handlers

ywangd committed:
checkstyle

ywangd committed:
Fix for no cache scenario

ywangd committed:
Merge remote-tracking branch 'origin/master' into es-53940-apikey-getdoc-avoidance

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityCacheRegistry.java:
```diff
@@ -0,0 +1,52 @@
+import static org.elasticsearch.xpack.security.support.SecurityIndexManager.isIndexDeleted;
+import static org.elasticsearch.xpack.security.support.SecurityIndexManager.isMoveFromRedToNonRed;
+
+public class SecurityCacheRegistry {
+
+    private static final Map<String, CacheInvalidator> cacheInvalidators = new ConcurrentHashMap<>();
```
I think we probably prefer to have a singleton instance for this class similar to all other components. If so, I can make it so and have it injected into the Transport action.

ywangd commented:
@tvernum After a few back and forth, here is the version that I am settled with. It touches the idea of consolidation for cache implementaitons, though with a much less ambitious approach. In addition to adding a cache for API key document, it also provides two enhancements:
* A [SecurityCache](https://github.com/elastic/elasticsearch/pull/59376/files#diff-ad9e28d1a2fc8243585e7f3216f081a1R74) class to encapsulate best practicies around caching expiry and stale results. It is fairly simple in that it does not try to be a high level generic solution to all cache requirements, which is difficult and the complexity is not worth it (I had a version with 5 type parameters). Thus it requires a bit of manual work when using it. We can improve it down the path, but for now I think it is reasonable start point.
* A [SecurityCacheRegistry](https://github.com/elastic/elasticsearch/pull/59376/files#diff-f2ba44027fdc231a7edb54696ff43ac9) that centralise the caching clearing logic for Rest/Transport handlers. Right now it is just used for API key caches, including the new API key doc cache and the existing auth cache. But I can see it to replace the handlers for role cache and privilege cache and possibly realm cache as well. These can be done in a separate PR if the approach here looks good to you. This possibility is also in line with your previous [comment](https://github.com/elastic/elasticsearch/pull/55836#discussion_r431610879) on the privilege cache PR.

The PR still needs a few more things:
* Tests
* Client side ecode
* Documentation

But I'd appreciate your review on the approach before I polish it.

ywangd committed:
Merge remote-tracking branch 'origin/master' into es-53940-apikey-getdoc-avoidance

ywangd committed:
Simplify security cache based on feedback.

ywangd commented:
@tvernum I updated the PR based on your suggestion with some name tweaks. The changes are basically to simplify the SecurityCache, which in fact is now replaced by [`ConsistentCache`](https://github.com/elastic/elasticsearch/pull/59376/files#diff-6c7d16b306e917f79896b6a1961b72fd). The `ConsistentCache#checkpoint` method returns a `Checkpoint` interface that can be used to cache given key/value pair while minimizing the possibility of caching stale results. 

The `SecurityCacheRegistry` is renamed to `CacheInvalidatorRegistry` since that is all it does, i.e. providing cache invalidation services to different caches.

Could you please reivew the current approach? I'll proceeds with adding tests etc once we can agree on the overall approach. Thanks!

ywangd committed:
Merge remote-tracking branch 'origin/master' into es-53940-apikey-getdoc-avoidance

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java:
```diff
@@ -192,14 +199,42 @@ public ApiKeyService(Settings settings, Clock clock, Client client, XPackLicense
+                .setMaximumWeight(maximumWeight)
                 .build();
+            this.roleDescriptorsBytesCache = CacheBuilder.<String, BytesReference>builder()
+                .setExpireAfterAccess(ttl)
```
I wonder about this.
It defaults to 24h, and we don 't have much in place to clean it up.
Should it really cache for that long? It's expiry after access, so 1h feels like it would be sufficient.

From my brief reading, once the doc cache expires (5m after last write) there will be no incoming references to these entries. They may get revived, but I don't think holding onto that memory for 24h is justified.

ywangd replied:
I changed it to 1h. It was set to 24h because
* 24h is an existing setting, so felt easy to reuse. I understand technically it has different meaning. But the subtlety here feels irrelevant to end-users.
* Unlike API key document cache, the role descriptor cache, in theory, never "expires" in that they will never be outdated. Sure the entries may no longer be needed, but the association between a hash and its original content can never be invalid. Also many API keys may share the same role descriptors, for example, the `superuser` (limiting) role. These made me feel it is plausible to cache the descriptors for longer. Because even when a Key is expired or invalidated or haven't checked in for a while, it is likey a new key will be replacing it and the new key will have the exact same role descriptors and thus reuse the cache entries.

With above being said, I don't think the benefit of caching for longer is anyway significant, since the main performance hit is the document retrieval itself. So I am OK to have it at a shorter duration.

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/CacheInvalidatorRegistry.java:
```diff
@@ -0,0 +1,55 @@
+        }
+    }
+
+    public static void invalidate(String name, Collection<String> keys) {
```
```suggestion
    public static void invalidateByKey(String cacheName, Collection<String> keys) {
```

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/CacheInvalidatorRegistry.java:
```diff
@@ -0,0 +1,55 @@
+        }
+    }
+
+    public static void invalidateAll(String name) {
```
This name implies invalidating _all_ caches, not all entries in 1 cache.

```suggestion
    public static void invalidateCache(String cacheName) {
```

ywangd committed:
Apply suggestions from code review

Co-authored-by: Tim Vernum <tim@adjective.org>

ywangd committed:
feedback

ywangd committed:
feedback

ywangd committed:
Merge remote-tracking branch 'origin/master' into es-53940-apikey-getdoc-avoidance

tvernum commented:
This approach looks good to me.

ywangd committed:
Merge remote-tracking branch 'origin/master' into es-53940-apikey-getdoc-avoidance

ywangd committed:
Add tests for api key doc caches

ywangd committed:
One more scenario for cache miss

ywangd committed:
Replace static class with the existing component pattern

ywangd committed:
Add tests for consistent cache

ywangd committed:
Add tests for cache invalidator registry

ywangd committed:
Add tests for integration tests for cache invalidation

ywangd committed:
Minor test tweak

ywangd committed:
Add client code, yaml tests and documentation

ywangd committed:
Fix json spec file

ywangd committed:
Fix tests

ywangd committed:
update doc

elasticmachine commented:
Pinging @elastic/es-security (:Security/Authentication)

ywangd commented:
@tvernum Just removed the "draft" status and this PR is now an official one. Here is a summary of new changes:

* Add client side code (HLRC) for cache clearing
* Add documentation changes for both server API and HLRC, previews are:
  - [HLRC](https://elasticsearch_59376.docs-preview.app.elstc.co/guide/en/elasticsearch/client/java-rest/master/java-rest-high-security-clear-security-cache.html)
  - [Server API](https://elasticsearch_59376.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/security-api-clear-security-cache.html)
* Tests added (java, yaml and doc tests)
* Replace the static class usage of `CacheInvalidatorRegistry` with component created inside `Security` class to be consistent with existing code pattern.

I'd like to migrate cache clearing of realm users, roles and application privileges to use the new classes added in this PR. But I think it deserves its own PR. So I'll address it later after this one is resolved. Thanks!

albertzaharovits commented:
@ywangd I've got a basic question.

If I understand it right, the root problem is that the existing API key cache sits **after** the doc-get operation?
I _think_ we can move the **existing** cache before the doc-get operation. The hash (SHA256) of the API key secret should be part of the cache key. A cache miss means that either the key id or secret are wrong. A cache hit must still verify the expiration.
I believe this is a simplification compared to the two caches, from this PR, one for the doc and the other for the authentication result. It also represents an improvement on the space savings, which is the being reason for the third cache.

I might be missing context here.

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/ConsistentCache.java:
```diff
@@ -0,0 +1,80 @@
+    public void invalidate(Collection<K> keys) {
+        try (ReleasableLock ignored = invalidationWriteLock.acquire()) {
+            numInvalidation.incrementAndGet();
+        }
```
do we really need a lock to protect an atomic variable? What does this achieve.

ywangd replied:
Please see my other [comment](https://github.com/elastic/elasticsearch/pull/59376#discussion_r498621504)

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java:
```diff
@@ -364,13 +425,16 @@ private void loadApiKeyAndValidateCredentials(ThreadContext ctx, ApiKeyCredentia
-                            listener.onFailure(e);
+                    if (checkpoint != null) {
+                        final CachedApiKeyDoc cachedApiKeyDoc = apiKeyDoc.toCachedApiKeyDoc();
+                        if (checkpoint.put(docId, cachedApiKeyDoc)) {
```
It's a bit hard to understand the reasoning for the "consistent cache" here.

My understanding is that you're trying to guard against a doc update (a call to invalidate a key), followed by a cache clear call (they are currently separate endpoints), **after** the doc has been retrieved by the get, but before the handler is called. IMO, this is a narrow time window for this big of a complication (the `ConsistentCache` complication).

Alternatively, what do you say you make the **current** invalidation counter part of the cached **value** (the `CachedApiKeyDoc` instance). In this case, you always populate the cache values with the current invalidation counter, but you never use a cache entry with an older invalidation counter compared to the current (you remove the entry and try the get).

ywangd replied:
The `ConsistentCache` is an attempt to extract a pattern from existing code so that caching behaviour can be reused in different places. The [`CompositeRolesStore`](https://github.com/elastic/elasticsearch/blob/7.9/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStore.java#L301-L312) is an example of this pattern, which is also applied for [`NativePrivilegeStore`](https://github.com/elastic/elasticsearch/blob/7.9/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStore.java#L432-L439) in my previous PR (#55836) for caching application privileges. The complexity is real but it is not new. It is exactly because of the complexity, I am attempting to extract the logic in a separate class so it can be reused, e.g. replacing the existing caching code of `NativePrivilegeStore` in future PRs.

The purpose of this piece of code is to minimize the possibility of caching stale results and invalidate it as soon as possible if we do end up caching it. Tim and I had some [discussion](https://github.com/elastic/elasticsearch/pull/55836#discussion_r416298343) about how it works. Unfortunately, I just realised that I didn't add a comment after our discussion so that the reasonsing is not obvious. I will try to capture it here. In summary, both the `AtomicLong` and `ReadWriteLock` are necessary due to the aysnc and non-atomic nature of operations in the system. The problem arises because of following few things:
* The read and write calls to the security index are independant. 
* The clear cache call is independant as well. This include both standalone clear cache call initiated by the users or the ones as consequence of write calls.
* `AtomicLong#get` and any actions guarded by it are not an atomic group.

The `ReadWriteLock` helps make above async stuff easier to reason about and minimize the possibility of stable results being cache. We can use following scenario as an example:

A read successfully retrieves an document from the security index. But before the code is done with retrieved document, an write call happens. The write call basically makes the last read document stale. So we do not want to cache the retrieved document. The write call fires a cache invalidation request, which is currently inflight. Before actuall caching the document, the read thread checks whether invalidation has happened with following code:
```java
if (invalidationCounter == numInvalidation.get()) {
    cacheTheDocument();
}
```
However, even if the check passes, it is still possible that the invalidation thread gets executed immedilatey after the check and before the actual caching call. So the read thread will end up caching a stale document. The chance for this happen may be low, but it is hard to reason about because it is a racing condition. When it happened, it is likely an API key is invalidated, but because it is errorneously cached, subsequent requests will only see the cached result.

When we add the `ReadWriteLock`, the code becomes:
```java
try (ReleasableLock ignored = invalidationReadLock.acquire()) {
    if (invalidationCounter == numInvalidation.get()) {
        cacheTheDocument();
    }
}
```
It helps with above issue because: If the read thread reaches the critical region first and acquires the lock, it can be sure that the counter will not suddenly change again when `cacheTheDocument` is executed. In another word, invalidation thread will be put on hold because of the locking. So even if the read thread does cache the stale document, once it releases the lock, invalidation thread will acquire it and invalidate the document momentarily. Thus minimises the possibility of caching stale document.

Witout the `ReadWriteLock`, it does not matter where you place the `AtomicLong` counter because the real problem is because anything guarded by the counter cannot be guarantteed to finish execution before the counter changes again. And this could happen regardless of where the code is.

albertzaharovits commented:
Apologies Yang, I've raised some pretty fundamental points, very late in the process (you have tests, etc) and after a long time since you've added me as a reviewer and a couple of days before FF. The worst 😢

ywangd commented:
> Apologies Yang, I've raised some pretty fundamental points, very late in the process (you have tests, etc) and after a long time since you've added me as a reviewer and a couple of days before FF. The worst 😢
>
No need to worry. They are good questions and I think it is in fact a good opportunity to capture some live discussions that Tim and I had (sorry that I didn't write down them earlier, time to review our communication charter ...)

> If I understand it right, the root problem is that the existing API key cache sits **after** the doc-get operation?
>
This is one of the problems, but not the root problem. The "root" problem is that the two caches, "document cache" and "authentication cache" have different lifetimes. The document cache has a changing nature because the document can be updated and it is also intentionally kept shortlived in case an invalidation call does not make it successfully. The authentication cache, on the other hand, has a longer lifetime because the hash for some secret does not change at all.

> I _think_ we can move the **existing** cache before the doc-get operation. The hash (SHA256) of the API key secret should be part of the cache key. A cache miss means that either the key id or secret are wrong. A cache hit must still verify the expiration.
> I believe this is a simplification compared to the two caches, from this PR, one for the doc and the other for the authentication result. It also represents an improvement on the space savings, which is the being reason for the third cache.
> 
Are you proposing have a single cache for both authentication and document instead of two? Or are you just suggesting their order to be swapped?

If it is the later, I don't see any benefits and it might actuall be more complicated, e.g. it is difficult to invalidate a wrong credentials when the correct one arrives.

If it is the former, i.e. single cache, I think it may help solving some issues, but also introduces its own problem and is not overall a better solution. I actually had an earlier version where a single cache is used for both of document and authentication. The difference is that I still kept using only the `docId` as the key. It worked but the code was very complicated because of the different lifetimes, which essentially requires a cache entry to be "partially" invalidated and updated. In addition, when an entry is being updated, we need to make sure no other thread will use the "partial" result. I abandonned it because I don't think it's worth the complication.

Using SHA value as part of the cache key is something that I have not considered before. I assume the cache key could be something like `Tuple(docId, SHA)`. It is a clever idea. But I think it has complications under the surface because of the lifetime issue as well. For example, when requests keep coming in with different wrong credentials of a new API key, they will end up fetching the document everytime and caching them repeatedly. Also when the correct credentials finally arrive, how do we clear the invalid entries? Since the cache is not keyed by just `docId`, it is difficult to keep track of all entries that have wrong credentials. Another example, assuming an API key is cached successfully with correct credentails and it now gets invalidated. We will face the same problem of "partial" update of a cache entry because the content has both document and "true/false" auth result. We could choose just invalidate the whole entry and reinsert it again. But when it is invalidated, another thread is probabaly gonna trying to insert it as well. In this case, how do we know which thread is correct (more recent)? There are other scenarios of similar natures. Overall the behaviour is harder to reason about even if it appears simpler on the surface.

Please let me know if my comments make sense. I am happy to have a live session if necessary. Just let me know. Thanks!

albertzaharovits commented:
@ywangd As discussed, here's my attempt to move the cache before the doc-get operation:
https://github.com/elastic/elasticsearch/compare/master...albertzaharovits:mock_api_cache_before_doc_get?expand=1

This is not trying to fix the concurrent invalidation issue (the `ConsistentCache`). It only illustrates that a single cache is sufficient, i.e. there doesn't have to be two, one before the doc-get, and another one before credential validation.

Note that, as is the case with this PR, a key invalidation must be followed by a clear cache API call (we could chain them server-side, but this is out of scope for this dummy PR).

I'll follow-up with my stab at concurrent invalidation.

ywangd commented:
> It only illustrates that a single cache is sufficient
> 
> Note that, as is the case with this PR, a key invalidation must be followed by a clear cache API call (we could chain them server-side, but this is out of scope for this dummy PR).
> 
If you let go some of the existing features of how this part of code works, it is possible to have a single cache. But then it becomes a different implementation with different promises and requires a different discussion. For examples:

* The mock code does not have negative caching. So a wrong API key ID or an expired one will trigger the more expensive process repetitively
* It also does not have `ListenableFuture` to avoid repeating expensive hash calculation
* It invalidates an expire key right away, so if the next request comes in with the same key, the code will have to retrieves the document again.
* It relies on a clear cache call to invalidate an API key if it is invalidated/updated. This by itself is not entirely reliable because a request can fail for many reasons and then leave the node in an inconsistent state for a long time. This also leads to the next point.
* It does not invalidate API key documents often enough, e.g. 5 min, and it cannot because it shares the same cache config as authentication result. We want it to be short-lived because we still believe fetching from the index is the most correct behaviour. But because fetching for every request is slow, the compromise is to fetch _a bit_ less often but not too much. 

One can argue that above features are not all that necessary. Some of them might be edge cases, some optimisations may not provide much gain, and some of them, again, is to minimise the possibility of stale results, i.e. node is out of sync. We can revisit them. However it then become a different discussion which is not what this PR attempts to address.

albertzaharovits commented:
I think your last point is the most important, and I've not considered it:

> We want it to be short-lived because we still believe fetching from the index is the most correct behaviour. But because fetching for every request is slow, the compromise is to fetch a bit less often but not too much.

Having two different cache expiration times for the get-doc and for the validate-credentials operations is a requirement that necessarily requires two caches; there's no way around it.

**I'm convinced by this.** I think the two operations can vary a lot in their duration, and also that we should be extra careful with the get-doc part which can become stale (unlike the other operation).

The other points are less convincing, for example the negative cache thing: neither the main nor this change contain a negative cache, so I don't consider its lack a detracting point on the part of my proposal.

ywangd commented:
> neither the main nor this change contain a negative cache
>
The existing code does have a negative cache. [This line](https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java#L559) caches the authentication result when `verified` is either `true` or `false`. If `verified` is a `false`, this entry becomes a negative caching and it is checked [here](https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java#L548-L549).

Each of the points may not be difficult to implement on its own using a single cache approach (even the last one is possible). But if you try to fulfill all of them together, it is likely the resulted code would become more complex than the 2-cache approach. I have a version [here](https://github.com/elastic/elasticsearch/compare/master...ywangd:es-53940-apikey-getdoc-avoidance-unified-cache#diff-cc4f9d4c564aae71214ac709ed789b4dR477-R599), which I didn't even bother asking for review.

albertzaharovits commented on x-pack/docs/en/rest-api/security/clear-security-cache.asciidoc:
```diff
@@ -0,0 +1,54 @@
+For clearing role cache, see the <<security-api-clear-role-cache,Clear roles cache API>>.
+For clearing privileges, see <<security-api-clear-privilege-cache,Clear privileges cache API>>.
+These separate APIs will be migrated over time.
+
```
This note here doesn't sound too "professional".
We have APIs and deprecated APIs , but no APIs that should do the job of others but not yet.
We can create an API now that we foresee will be deprecated. There is no rush, people can sit on a version for a long time.
Not a big deal, just my thoughts.

ywangd replied:
I will drop it and document it in the context of API key. See my [other comment](https://github.com/elastic/elasticsearch/pull/59376#discussion_r499199071).

albertzaharovits commented on client/rest-high-level/src/main/java/org/elasticsearch/client/security/ClearSecurityCacheResponse.java:
```diff
@@ -0,0 +1,50 @@
+/**
+ * The response object that will be returned when clearing a security cache
+ */
+public final class ClearSecurityCacheResponse extends SecurityNodesResponse {
```
Too bad this doesn't return the failed nodes. Not something to tackle atm.

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java:
```diff
@@ -364,13 +425,16 @@ private void loadApiKeyAndValidateCredentials(ThreadContext ctx, ApiKeyCredentia
+                            roleDescriptorsBytesCache.computeIfAbsent(
+                                cachedApiKeyDoc.roleDescriptorsHash, k -> apiKeyDoc.roleDescriptorsBytes);
+                            roleDescriptorsBytesCache.computeIfAbsent(
+                                cachedApiKeyDoc.limitedByRoleDescriptorsHash, k -> apiKeyDoc.limitedByRoleDescriptorsBytes);
```
As a possible refactoring in the future, the logic that puts and gets an API key doc, which uses a secondary cache for the role descriptors is a nice thing to encapsulate.

ywangd replied:
Add a new ApiKeyDocCache inner class to encapsulate the secondary cache. It is the right thing to do for this PR. Thanks!

albertzaharovits reviewed (commented):
Thank you Yang for the patience!
I now understand all the core inner working and I believe they work as expected.
But as explained in the comments, I'm not onboard with the general API to clear the caches; it looks too "low level" to me.
Also I think the `ConsistenCache` is the wrong abstraction, it doesn't help me abstract away anything, I have to read the code behind every method.

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/ConsistentCache.java:
```diff
@@ -0,0 +1,80 @@
+/**
+ * A wrapper of {@link Cache} that minimizes the possibility of caching stale results.
+ */
+public class ConsistentCache<K, V> {
```
I like this version. I will update the code like it except that I might change the class name to something like `StaleMinimizingCacheWrapper` to be more meaningful and specific. 

I agree that the ConsistentCache is not a good abstraction in that its interface is unintuitive to use. The background is that I started from a very complex version where I tried to provide an unified interface for security related caches and hide all implementation details, which turned out to be too much to hide. I have since tried to decompose it to a simpler version, i.e. the ConsistentCache. The general approach for simplication is to make inner steps explicit instead of trying to hide them (this is also Tim's suggestion) and also cater for fewer use cases. But my thoughts were still biased by the initial complex version, which prevented me to further simplfy it. For this, I think "pulling the counter out explicity" is indeed the right move. So the overall journey has been from an attempt to have "one cache to rule them all" to a more pragmatic one that is just for code reuse. It is less satisfying that I was not able to achieve the initial goal. But given what we have, I think the simpler abstraction is the right one for the time being.

albertzaharovits replied:
Not to dwell too much on it, but `StaleMinimizing...` is not what the class does but what it is used for. You're tying the class to a particular circumstance. It is "stale minimizing" only in the particular context that it is used in. I think it would be much simpler to explain how it achieves the minimization in a comment at the use site, like in the `CompositeRolesStore` class, than in the class's javadoc. The name is also a bit intimidating, IMO.

My name is inspired by `CacheIteratorHelper` .

ywangd replied:
One main reason that I don't think `CacheInvalidationHelper` is a proper name is because the class is essentially a `Cache`. If you look past the counter and lock, it is just a regular Cache. So I'd much prefer to call it something along the line of `XxxCache` or `XxxCacheWrapper`. "Minimizing the chane of stale entries" is the new feature added onto the exiting `Cache` class. Often times, classes like this are named as `AdjectiveNoun`, e.g. `Realm` and `CachingRealm`. So I think `StaleMinimizingCache[Wrapper]` is an accurate name. If you feel strong about `StaleMinimizing`, how about `InvalidationCountingCache[Wrapper]`? It is still accurate, but less comprehensive.

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java:
```diff
@@ -192,14 +199,42 @@ public ApiKeyService(Settings settings, Clock clock, Client client, XPackLicense
+        final Integer maximumWeight = CACHE_MAX_KEYS_SETTING.get(settings);
         if (ttl.getNanos() > 0) {
             this.apiKeyAuthCache = CacheBuilder.<String, ListenableFuture<CachedApiKeyHashResult>>builder()
                 .setExpireAfterWrite(ttl)
```
We already have "expire after xxx" for all caches. The maximum weight is to control the overall size of a cache so that it does not grow unbounded. They are for different purposes.

albertzaharovits replied:
I think I haven't expressed myself clearly, what I meant:
```suggestion
                .setExpireAfterAccess(ttl)
```
I think the "expire after write" is suitable for the doc cache but the credential cache might benefit more from the "expire after access" strategy because the underlying data (the key secret) never changes.

ywangd replied:
Ah OK I see your point now. This is existing code. I thought there was a reason why "expire after write" is used, but now I cannot remember any. I agree that "expire after access" seems to be a better fit. But I seem to remember noticing this previously, but I am not sure now why it was kept like this. I could be missing something. If you don't mind. I'll leave this out as future work.

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/RestClearSecurityCacheAction.java:
```diff
@@ -0,0 +1,47 @@
+
+    @Override
+    public List<Route> routes() {
+        return Collections.singletonList(new Route(POST, "/_security/{cache_name}/{keys}/_clear_cache"));
```
Sorry this is a mistake on my side and thanks for catching it. My intention for the new Clear Cache API is just an internal refactor and should not require user awareness. This is why I kept the Rest end point to be compatible with the existing ones, e.g. `/_security/privilege/{application}/_clear_cache`, where `privilege` is the cache name and `{application}` is the keys. My plan was to introduce the new cache clearing API as if it is just for API key, i.e, I planned to document it just like all the other existing APIs and what user should be seeing is just `/_security/api_key/{id}/_clear_cache`.

But somehow I got carried away by the changes and added documentation for the generic API itself, which like you said, is low level and not user friendly. Hence I will drop this documentation and just document it in the context of API key. Whether these cache clearing APIs internally use a single or different actions are implementation details and should not concern users.

albertzaharovits replied:
> On the surface (documentation), user will just see that we add another cache clearing API for API keys, i.e. /_security/api_key/{id}/_clear_cache.

The `_security/api_key` APIs have the parameters in the request body because of the multiple ways to refer to keys (by `id`, `name`, `user` and `realm`). If you put in the `id` in the URL path for the cache clear operation, it would be unlike the other APIs and would be more complex to introduce clearing of the cache for keys by other params.

ywangd replied:
I don't think we want to match the cache clearing API with other "api_key" APIs. It is never a goal to clear cache by API key names, users or realms. Caches are always cleared by the keys, i.e. `id` in API key's case.  So I don't see any problem here.

ywangd replied:
@albertzaharovits Per our discussion, I have [changed](https://github.com/elastic/elasticsearch/pull/59376/commits/11e591babca2733bc1ce37c692989e38c9586350#diff-f0edbeeb832bf03f92fffc111d46fbd2) the Rest Action class to be dedicated to "clear API key cache". Thanks for the feedback!

ywangd commented:
> I'm not onboard with the general API to clear the caches; it looks too "low level" to me.
>
> Also I think the `ConsistenCache` is the wrong abstraction, it doesn't help me abstract away anything, I have to read the code behind every method.
>
I think both are valuable inputs and I will address them. Detailed answers can be found [here](https://github.com/elastic/elasticsearch/pull/59376#discussion_r499199071) for the first item and [here](https://github.com/elastic/elasticsearch/pull/59376#discussion_r499198558) for the second item.

ywangd commented on client/rest-high-level/src/main/java/org/elasticsearch/client/security/ClearSecurityCacheRequest.java:
```diff
@@ -0,0 +1,75 @@
+ */
+public final class ClearSecurityCacheRequest implements Validatable {
+
+    private final String cacheName;
```
I think string is fine since there is a `CacheInvalidator` that already controls the available/valid entries. I don't see what extra value enum could provide here. Please also see my other [comment](https://github.com/elastic/elasticsearch/pull/59376#discussion_r499199071) about why it is a request param.

albertzaharovits replied:
The construction of `ClearSecurityCacheRequest` object happens on the client side. We might have validation for its parameters on the server side, but the caller has to hard code the "api_key" value in the client code. Something like:
```
public final class ClearSecurityCacheRequest implements Validatable {
    
    enum CacheName {
        API_KEY
    }

    private final CacheName cacheName;
```
would help a bit.

A separate request object type would still be preferable IMO.

ywangd replied:
Client side is a good point. Until we can have auto-generation of client from API spec, we'll have to maintain this manually. I added a dedicated `ClearApiKeyCacheRequest` for the client code.

ywangd commented:
I just realised that I forgot to add the critical cache clearing call after API key invalidation! Apparently no tests for this scenario as well. Will add them 🤦

ywangd committed:
Address feedback: simplify consistent cache

ywangd committed:
Address feedback: only expose clear api key cache API. Client side dedicate clear api key cache request class

ywangd committed:
Add cache clear call after key invalidation

ywangd committed:
checkstyle

ywangd committed:
More check style

ywangd commented:
@albertzaharovits I updated this PR per your review and discussion. It is ready for another look. Thanks!

ywangd committed:
Fix tests

ywangd committed:
Refactor to better manage the secondary role descriptor cache

ywangd committed:
null check optimize

ywangd committed:
Fix tests

ywangd committed:
checkstyle

ywangd committed:
Fix tests

ywangd committed:
Address feedback - add dedicated rest action for clear api key cache

ywangd committed:
Merge remote-tracking branch 'origin/master' into es-53940-apikey-getdoc-avoidance

ywangd committed:
Remove clear security request base class from client code

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java:
```diff
@@ -160,6 +168,8 @@
     public static final Setting<Integer> CACHE_MAX_KEYS_SETTING = Setting.intSetting("xpack.security.authc.api_key.cache.max_keys",
         10000, Property.NodeScope);
+    public static final Setting<TimeValue> DOC_CACHE_TTL_SETTING = Setting.timeSetting("xpack.security.authc.api_key.doc_cache.ttl",
+        TimeValue.timeValueMinutes(5), TimeValue.timeValueMinutes(0), TimeValue.timeValueMinutes(5), Property.NodeScope);
```
Why is the max 5min? Are we sure we don't ever want to let someone increase that TTL?
It seems like something we'd like to have available if we need to boost performance.
```suggestion
        TimeValue.timeValueMinutes(5), TimeValue.timeValueMinutes(0), TimeValue.timeValueMinutes(15), Property.NodeScope);
```

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java:
```diff
@@ -904,6 +965,25 @@ private void indexInvalidation(Collection<String> apiKeyIds, ActionListener<Inva
+                public void onFailure(Exception e) {
+                    logger.error("unable to clear API key cache", e);
+                    listener.onFailure(new ElasticsearchException(
+                        "clearing the API key cache failed. " + "please clear the caches manually", e));
```
```suggestion
                        "clearing the API key cache failed; please clear the caches manually", e));
```

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java:
```diff
@@ -1088,8 +1168,128 @@ public ApiKeyDoc(
+
+    /**
+     * A cached version of the {@link ApiKeyDoc}. The main difference is that the role descriptors
+     * are replaced by their hashes. The actual values are stored in a separate role descriptor cache.
```
```suggestion
     * are replaced by their hashes. The actual values are stored in a separate role descriptor cache,
     * so that duplicate role descriptors are cached only once (and therefore consume less memory).
```

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java:
```diff
@@ -1088,8 +1168,128 @@ public ApiKeyDoc(
+                    .setExpireAfterWrite(ttl)
+                    .build()
+            );
+            this.roleDescriptorsBytesCache = CacheBuilder.<String, BytesReference>builder()
```
```suggestion
            // We don't use the doc TTL because that TTL is very low to avoid the risk of
            // caching an invalidated API key. But role descriptors are immutable and may be shared between
            // multiple API keys, so we cache for longer and rely on the weight to manage the cache size.
            this.roleDescriptorsBytesCache = CacheBuilder.<String, BytesReference>builder()
```

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/TransportClearSecurityCacheAction.java:
```diff
@@ -0,0 +1,76 @@
+import java.io.IOException;
+import java.util.List;
+
+public class TransportClearSecurityCacheAction extends TransportNodesAction<ClearSecurityCacheRequest, ClearSecurityCacheResponse,
```
```suggestion
/**
 * Clears a security cache by name (with optional keys).
 * @see CacheInvalidatorRegistry
 */
public class TransportClearSecurityCacheAction extends TransportNodesAction<ClearSecurityCacheRequest, ClearSecurityCacheResponse,
```

tvernum commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/ClearSecurityCacheRequest.java:
```diff
@@ -0,0 +1,86 @@
+        super.writeTo(out);
+        out.writeString(cacheName);
+        out.writeOptionalStringArray(keys);
+    }
```
Ideally we would have serialization tests for every class that has a `writeTo` method.

ywangd replied:
Add a serialisation [test](https://github.com/elastic/elasticsearch/pull/59376/commits/ef95a1dd11410530bccba9d0ddbbb46d50ea4e82#diff-675acc3ee8baa4e58b409727e9b8958c).

tvernum commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/support/CacheInvalidatorRegistryTests.java:
```diff
@@ -0,0 +1,79 @@
+        verify(invalidator2).invalidateAll();
+    }
+
+    public void testInvalidateByKey() {
```
A Nit, with a real implication.

This test name just refers to the method that it tests. It doesn't actually say what behaviour it's testing, just the method it calls. A name like
```suggestion
    public void testInvalidateByKeyCallsCorrectInvalidatorObject() {
```
would be more descriptive.

_And_ the implication is that because we have a method named `test{SomeMethod}` it leads to thinking that we just need to call some stuff on that method and we're done, and skips the thinking step of _What cases do I need to test?_.  
And in this case, there's no tests for invalidating a cache that hasn't been registered. Should that be a no-op? Or a failure?

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/CacheInvalidatorRegistry.java:
```diff
@@ -0,0 +1,60 @@
+        final CacheInvalidator cacheInvalidator = cacheInvalidators.get(cacheName);
+        if (cacheInvalidator != null) {
+            cacheInvalidator.invalidate(keys);
+        }
```
The lack of an `else` seems like dangerous leniency. Why do we silently pretend to invalidate a cache that doesn't exist?

ywangd replied:
This is a good point and there is no reason for this leniency. Add an [else branch](https://github.com/elastic/elasticsearch/pull/59376/commits/ef95a1dd11410530bccba9d0ddbbb46d50ea4e82#diff-95381cf595a43e010005cf0de5538f56) to throw exception and also relevant tests.

tvernum commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/support/InvalidationCountingCacheWrapperTests.java:
```diff
@@ -0,0 +1,70 @@
+        assertEquals("bar", invalidationCountingCacheWrapper.get("foo"));
+    }
+
+    public void testItemWillNotBeCachedWhenInvalidationHappensBetweenCheckpointAndItsUsage() throws InterruptedException {
```
```suggestion
    public void testItemWillNotBeCachedIfInvalidationCounterHasChanged() throws InterruptedException {
```

tvernum reviewed (approved):
LGTM, but I have a few suggestions (some of which are strong suggestions).

ywangd committed:
Apply suggestions from code review

Co-authored-by: Tim Vernum <tim@adjective.org>

ywangd committed:
Apply suggestions from code review

Co-authored-by: Tim Vernum <tim@adjective.org>

ywangd committed:
Address feedback

albertzaharovits reviewed (approved):
LGTM Good efforts on this one!
